### ISSUE TITLE
torch.bernoulli inconsistent gpu/cpu results

### ISSUE THREAD

**Issue by davidmascharka** [pytorch/pytorch#10357]
## Issue description

The behavior of CUDA and CPU `sample` in the `distributions` module is inconsistent. On the CPU side, there appears to be an argument validity check, while on the GPU side, this does not appear to be implemented. [This issue](https://github.com/pytorch/pytorch/issues/5248) is somewhat related.

## Code example

```python
>>> torch.distributions.Bernoulli(torch.FloatTensor([1.5])).sample()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-150-d17a827b2cc3> in <module>()
----> 1 torch.distributions.Bernoulli(torch.FloatTensor([1.5])).sample()

python3.6/site-packages/torch/distributions/bernoulli.py in sample(self, sample_shape)
     72         shape = self._extended_shape(sample_shape)
     73         with torch.no_grad():
---> 74             return torch.bernoulli(self.probs.expand(shape))
     75 
     76     def log_prob(self, value):

RuntimeError: invalid argument 1: must be >= 0 and <= 1 at /pytorch/aten/src/TH/THRandom.cpp:314

>>> torch.distributions.Bernoulli(torch.cuda.FloatTensor([1.5])).sample()
tensor([1.], device='cuda:0')

>>> torch.distributions.Bernoulli(torch.cuda.FloatTensor([-2])).sample()
tensor([0.], device='cuda:0')
```

## System Info

```
PyTorch version: 0.5.0a0+a24163a
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.9.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 390.67
cuDNN version: 7

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.7.0)
[pip] torch (0.5.0a0+a24163a)
[pip] torchvision (0.1.9)
```

**Comment by vishwakftw** [comment]
Distributions have a `validate_args` argument that does this for you. The default value is `False`. Could you please set it to true, and check please.

**Comment by vishwakftw** [comment]
I believe you should pass `validate_args` in the constructor of the distribution.

**Comment by davidmascharka** [comment]
That may be recommended. However, that is not the issue. The issue is the inconsistency of the `sample()` function between the CUDA and the CPU implementations. Having the distribution's `__init__` function throw is completely beside the point here. The fact is I *can* create a distribution with `validate_args` set to `False` or `None` in both the CPU and the CUDA versions. However, the `sample()` functions then differ in their behaviour. They should behave exactly the same.

**Comment by vishwakftw** [comment]
cc: @fritzo

**Comment by fritzo** [comment]
@davidmascharka I believe you are hitting bug #9917 in `torch.bernoulli(p.expand())`. The [workaround](https://github.com/uber/pyro/pull/1294/files#diff-120fbe36e4b03a2335f19d622512e077R51) we're using in Pyro is to call `.contiguous()` on the argument to `torch.bernoulli()`:
```py
torch.bernoulli(p.expand(shape))               # broken in 0.4.1
torch.bernoulli(p.expand(shape).contiguous())  # workaround
```

**Comment by davidmascharka** [comment]
@fritzo that actually points me to a problem with just `torch.bernoulli`:

``` python
>>>  p = torch.FloatTensor([1.5])

>>> torch.bernoulli(p) # note no expand
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-10ad530fb7ed> in <module>()
----> 1 torch.bernoulli(p)

RuntimeError: invalid argument 1: must be >= 0 and <= 1 at /pytorch/aten/src/TH/THRandom.cpp:314


>>> torch.bernoulli(p.cuda())
tensor([1.], device='cuda:0')
```

**Comment by SsnL** [comment]
This should eventually be fixed by my patch #10273 .

**Comment by SsnL** [comment]
fixed in  https://github.com/pytorch/pytorch/pull/10273
